x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ScoringLuceneTopNSourceOperator.java

```diff
@@ -0,0 +1,156 @@
+            int limit,
+            List<SortBuilder<?>> sorts
+        ) {
+            super(contexts, queryFunction, dataPartitioning, taskConcurrency, maxPageSize, limit, sorts, ScoreMode.TOP_DOCS_WITH_SCORES);
+        }
+
+        @Override
+        public SourceOperator get(DriverContext driverContext) {
+            return new ScoringLuceneTopNSourceOperator(driverContext.blockFactory(), maxPageSize, sorts, limit, sliceQueue);
+        }
+
+        @Override
+        public String describe() {
+            String notPrettySorts = sorts.stream().map(Strings::toString).collect(Collectors.joining(","));
+            return "ScoringLuceneTopNSourceOperator[dataPartitioning = "
+                + dataPartitioning
+                + ", maxPageSize = "
+                + maxPageSize
+                + ", limit = "
+                + limit
+                + ", sorts = ["
+                + notPrettySorts
+                + "]]";
+        }
+    }
+
+    public ScoringLuceneTopNSourceOperator(
+        BlockFactory blockFactory,
+        int maxPageSize,
+        List<SortBuilder<?>> sorts,
+        int limit,
+        LuceneSliceQueue sliceQueue
+    ) {
+        super(blockFactory, maxPageSize, sorts, limit, sliceQueue);
+    }
+
+    @Override
+    protected DoubleVector.Builder scoreVectorOrNull(int size) {
+        return blockFactory.newDoubleVectorFixedBuilder(size);
+    }
+
+    @Override
+    protected void consumeScore(ScoreDoc scoreDoc, DoubleVector.Builder currentScoresBuilder) {
+        if (currentScoresBuilder != null) {
+            float score = getScore(scoreDoc);
+            currentScoresBuilder.appendDouble(score);
+        }
+    }
+
+    @Override
+    protected Page maybeAppendScore(Page page, DoubleVector.Builder currentScoresBuilder) {
+        return page.appendBlocks(new Block[] { currentScoresBuilder.build().asBlock() });
+    }
+
+    float getScore(ScoreDoc scoreDoc) {
+        FieldDoc fieldDoc = (FieldDoc) scoreDoc;
+        if (Float.isNaN(fieldDoc.score)) {
+            if (sorts != null) {
+                return (Float) fieldDoc.fields[sorts.size() + 1];
+            } else {
+                return (Float) fieldDoc.fields[1];
+            }
+        } else {
+            return fieldDoc.score;
+        }
+    }
+
+    @Override
+    PerShardCollector newPerShardCollector(ShardContext shardContext, List<SortBuilder<?>> sorts, int limit) throws IOException {
+        Optional<SortAndFormats> sortAndFormats = shardContext.buildSort(sorts);
+        Sort sort;
+        if (sortAndFormats.isPresent()) {
+            var l = new ArrayList<>(Arrays.asList(sortAndFormats.get().sort.getSort()));
+            l.add(SortField.FIELD_DOC);
+            l.add(SortField.FIELD_SCORE);
+            sort = new Sort(l.toArray(SortField[]::new));
+        } else {
+            sort = new Sort(SortField.FIELD_DOC, SortField.FIELD_SCORE);
+        }
+        return new ScoringPerShardCollector(shardContext, sort, limit);
```

Comment:
@tteofili, do we want to not score at all?
Generally, TopN is limit + sort. I think we should have a _score extraction (also?) outside TopN, which is limit+sort.

Reply:
currently when adding `METADATA _score`, the `_score` is always extracted, even when no SORT is specified, is this what you were referring to? Sorry, I am not sure I got the point here @bpintea

Reply:
> _score is always extracted, even when no SORT is specified

When no `SORT` is explicitly specified, the `Sort` object built at :128 is used. This should not be the case and in general, if no `SORT` is provided, we should not end up in a TopN flow. 

I _assume_ we can extract the `_score` without sorting?

Reply:
> When no SORT is explicitly specified, the Sort object built at :128 is used. This should not be the case and in general, if no SORT is provided, we should not end up in a TopN flow.

the code at line 128 specified to sort by FIELD_DOC first (sort by docId aka physical ordering) is what happens behind the curtains in Lucene when no scoring is used. Adding a secondary sort by score is what triggers the collection of scores (as if they were field values), this is all Lucene internal logic. Note that the way it's implemented right now is almost identical to how it's done when [track_scores](https://www.elastic.co/guide/en/elasticsearch/reference/current/sort-search-results.html#_track_scores) is enabled. 
What we may do (perhaps in a followup PR) is implementing a custom `Collector` / `BulkScorer` to eventually make this easier to read (_disregarding_ the `Sort` object in case no sort option is specified).

Reply:
Sorry @tteofili, I'm not having much expertise in this area, so feel free to ignore my review.

> adding a secondary sort by score is what triggers the collection of scores

So, you're then saying that getting the score without sorting (within Lucene, not explicitly ES|QL) is not possible?

Looking at how `LuceneSourceOperator.Factory` creates the `LuceneSliceQueue` using a `ScoreMode.COMPLETE_NO_SCORES` and seeing there's an alternative `ScoreMode.COMPLETE` which is init'ed with `needsScore` of `true`, I'm wondering if `TOP_xxx` score mode is actually needed, if the user doesn't sort.
In other words, I'm wondering if we need to sort to get the score.

> Note that the way it's implemented right now is almost identical to how it's done when [track_scores](https://www.elastic.co/guide/en/elasticsearch/reference/current/sort-search-results.html#_track_scores) is enabled. 

But that reads `When sorting on a field, scores are not computed.`. If I'm not providing a sorting, scores are still computed (with `_search`).

Reply:
I think I see your point better now. I've tried to address your comment in [1230eb5](https://github.com/elastic/elasticsearch/pull/113120/commits/1230eb5e48796c02b7c658bd6f4920fd10ea5249).
Technically what happens in Lucene is the same, but I agree that with `ScoreMode.COMPLETE` and without an explicit `Sort` when sorting is not specified makes is a bit cleaner, thanks !

Reply:
I've added some more refactoring in [6abba22](https://github.com/elastic/elasticsearch/pull/113120/commits/6abba2208a2cd116f664267305acf18434af41a5) to use explicitly different `Collectors` depending on whether sorting is requested and / or if it's by score or some other fields.

Reply:
Great, thanks, I think this is going in the right direction.

Can we also then move the extraction out of TopN, when no sorting is applied? I think we can do it in this PR, since even if it works like this, it's not the right solution. Otoh, it won't enlarge it significantly and it's not time sensitive, I believe.

Alternatively, can we have an issue with next steps to track follow-ups of this PR?

Reply:
@bpintea @tteofili this is incomplete, butI believe that it addresses the comments raised here. https://github.com/tteofili/elasticsearch/pull/1.  ( it attempts to fold scoring into existing Lucene operator, rather than adding a new scoring operator ). That way scoring is a property of all lucene source operators, whether enabled or not it actually has little effect on the code flow. Additionally, the topN operator is only used when sorting is enabled.

Reply:
thanks @ChrisHegarty I've pushed the changes from your PR :)